### DISCUSSION TITLE
使用 CMake 编译 POSIX 版本，NDK 提示 cannot find -lpthread，是否需要 link pthread？

### DISCUSSION THREAD

**Discussion by yysongzy** [Tencent/MMKV#979] [Q&A]
```shell
Linking CXX shared library libmmkv.so
~/Library/Android/sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/darwin-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

```cmake
target_link_libraries( # Specifies the target library.
                       mmkv

                       # Links the target library to the log library
                       # included in the NDK.
                       #          ${log-lib}
                       core
        pthread
        )
```
编译安卓下的 POSIX 版本，NDK 报这个错，是否需要 link pthread？

**Comment by lingol** [comment]
Yes, `pthread` is needed.

**Comment by lingol** [comment]
What you should do is Google "ndk cannot find pthread" instead of complaining about it here.

**Reply by yysongzy** [reply]
The answer I got from Google is that the pthread is unnecessary for NDK, due to the Android libc and bionic provide built-in support for pthread.
Please refer to these three links.
https://stackoverflow.com/questions/38666609/cant-find-lpthread-when-cross-compile-to-arm
https://stackoverflow.com/questions/11048260/does-android-support-thread
https://stackoverflow.com/questions/57289494/ndk-r20-ld-ld-error-cannot-find-lpthread

**Reply by yysongzy** [reply]
> Note that on Android, unlike Linux, there are no separate libpthread or librt libraries. That functionality is included directly in libc, which does not need to be explicitly linked against.

As we can see from the [official guides](https://developer.android.google.cn/ndk/guides/stable_apis), we needn't to link pthread explicitly.

**Comment by lingol** [comment]
It's none of my business. Complain it to Android or NDK.
